src/libPMacc/include/particles/ParticlesBase.kernel

```diff
@@ -392,26 +403,27 @@ __global__ void kernelFillGaps(ParticlesBox<FRAME, Mapping::Dim> pb, Mapping map
         __syncthreads();
         if (threadIdx.x == 0)
         {
+            bool isFrameValid = false;
             if (counterGaps < counterParticles)
             {
                 //any gap in the first frame is filled
-                firstFrame = &(pb.getNextFrame(*firstFrame, isValid));
+                firstFrame = &(pb.getNextFrame(*firstFrame, isFrameValid));
             }
             else if (counterGaps > counterParticles)
             {
                 //we need more particles
-                lastFrame = &(pb.getPreviousFrame(*lastFrame, isValid));
-                isValid = isValid && pb.removeLastFrame(superCellIdx);
+                isFrameValid = getPreviousFrameAndRemoveLastFrame(lastFrame, pb, superCellIdx);
             }
             else if (counterGaps == counterParticles)
             {
-                lastFrame = &(pb.getPreviousFrame(*lastFrame, isValid));
-                isValid = isValid && pb.removeLastFrame(superCellIdx);
-                if (isValid && lastFrame != firstFrame)
+                isFrameValid = getPreviousFrameAndRemoveLastFrame(lastFrame, pb, superCellIdx);
+                if (isFrameValid && lastFrame != firstFrame)
                 {
-                    firstFrame = &(pb.getNextFrame(*firstFrame, isValid));
+                    firstFrame = &(pb.getNextFrame(*firstFrame, isFrameValid));
```

Comment:
some vals like `firstFrame` are manipulated here but until the next `while` loop there is no `__syncthreads();` called - sure that is still correct? looks like a race condition to me.